### PR TITLE
Add extraction of sub-ingredients and computation of possible percent ranges

### PR DESCRIPTION
**Pull request by stephanegigandet**
This PR contains 2 changes related to ingredients parsing:

- ingredients parsing now creates a nested array for sub_ingredients in the ingredients field.
see bug #2855 for details

- we now compute minimum and maximum percent values for all ingredients, based on known values.
see bug #2793 for details

Also added some perpod doc + tests related to those 2 features.

### CODE REVIEW

File: t/ingredients_tags.t [hunk 1]
```diff
@@ -82,6 +82,8 @@ foreach my $test_ref (@tests) {
 	my $product_ref = $test_ref->[0];
 	my $expected_tags = $test_ref->[1];
 
+	print STDERR "ingredients_text: " . $product_ref->{ingredients_text} . "\n";
```
Comment (hangy) [inline t/ingredients_tags.t]: `print STDERR` isn't recommended in unit tests: https://stackoverflow.com/a/6162805/11963

File: t/ingredients_percent.t [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+        ],
+
+);
+
+foreach my $test_ref (@tests) {
+
+	my $product_ref = $test_ref->[0];
+	my $expected_ingredients_ref = $test_ref->[1];
+
+	print STDERR "ingredients_text: " . $product_ref->{ingredients_text} . "\n";
```
Comment (hangy) [inline t/ingredients_percent.t]: `print STDERR` isn't recommended in unit tests: https://stackoverflow.com/a/6162805/11963

File: t/ingredients_percent.t [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+	compute_ingredients_percent_values(
+		100, 100, $product_ref->{ingredients});
+
+	is_deeply ($product_ref->{ingredients}, $expected_ingredients_ref)
+		# using print + join instead of diag so that we don't have
+		# hashtags. It makes copy/pasting the resulting structure
+		# inside the test file much easier when tests results need
+		# to be updated. Caveat is that it might interfere with
+		# test output.
+		or print STDERR join("\n", explain $product_ref->{ingredients});
```
Comment (hangy) [inline t/ingredients_percent.t]: `print STDERR` isn't recommended in unit tests: https://stackoverflow.com/a/6162805/11963

File: t/ingredients_nesting.t [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	],
+
+);
+
+foreach my $test_ref (@tests) {
+
+	my $product_ref = $test_ref->[0];
+	my $expected_ingredients_ref = $test_ref->[1];
+
+	print STDERR "ingredients_text: " . $product_ref->{ingredients_text} . "\n";
```
Comment (hangy) [inline t/ingredients_nesting.t]: `print STDERR` isn't recommended in unit tests: https://stackoverflow.com/a/6162805/11963

File: t/ingredients_nesting.t [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+	parse_ingredients_text($product_ref);
+
+	is_deeply ($product_ref->{ingredients}, $expected_ingredients_ref)
+		# using print + join instead of diag so that we don't have
+		# hashtags. It makes copy/pasting the resulting structure
+		# inside the test file much easier when tests results need
+		# to be updated. Caveat is that it might interfere with
+		# test output.
+		or print STDERR join("\n", explain $product_ref->{ingredients});
```
Comment (hangy) [inline t/ingredients_nesting.t]: `print STDERR` isn't recommended in unit tests: https://stackoverflow.com/a/6162805/11963

File: lib/ProductOpener/Ingredients.pm [hunk 6]
```diff
@@ -634,26 +681,33 @@ sub extract_ingredients_from_text($) {
 
 			if ($sep =~ /(:|\[|\{|\()/i) {
 
+				# Single separators like commas and dashes
+				my $match = '.*?';	# non greedy match
 				my $ending = $last_separator;
 				if (not defined $ending) {
 					$ending = ",|;|( $dashes )";
```
Comment (hangy) [inline lib/ProductOpener/Ingredients.pm]: Would it make sense to support `$commas` here? We had other issues where stuff like full-width/chinese commas weren't supported correctly in taxonomy fields

Comment (stephanegigandet) [reply lib/ProductOpener/Ingredients.pm]: Sure, I changed it to $commas

**Pull request merged**